네트워크 통신(HTTP) 구현

HJEHA opened the pull request:
### 구현 내용
- NetworkService : 네트워트 통신을 위한 구조체
  - loadData() : task를 만들고 completionHandler로 Response을 가공하는 메서드
  - request() : 요청을 보내는 메서드
- NetworkError : 네트워크 통신 에러 열거형
- Gettable Protocol : 기능별(Ticker, Orderbook 등) API(GET) 타입을 생성할 때 채택하는 프로토콜
   -  APIProtocol을 상속받음
- URLRequest+extension
  - APIProtocol을 매개변수로 받는 init?() 확장
- URLSession+extension
  - 실제 네트워크 통신에 의존하지 않는 테스트를 위한 URLSessionProtocol 채택
 
- MockURLSession
  - 실제 네트워크 통신에 의존하지 않는 테스트를 위한 타입
- NetworkServiceTests
  - URL이 유효하지않을때 statusCodeError가 나오는지
  - URL이 Nil일때 invalidURLRequest에러가 나오는지
  - MockURLSession의 isSuccess가 true일때 정상동작 하는지
  - MockURLSession의 isSuccess가 false일때 실패하는지
  - MockURLSession의 MockError을 담아보냈을때 MockError을 반환하는지
  Unit Test 진행
  - Unit Test 결과
    <img width="356" alt="스크린샷 2022-02-24 오후 6 06 41" src="https://user-images.githubusercontent.com/98801129/155493285-384778ff-4a9c-484b-8c1d-ccccbaea4f09.png">

Commit:
feat: NetworkServiec 타입 기본 구현 #3

- session 프로토콜 추가
- 이니셜라이져 추가

Commit:
feat: NetworkService loadData() 구현 #3

- NetworkError 열거형 구현

Commit:
feat: NetworkService request() 구현 #3

Commit:
feat: APIProtocol 구현 #3

Commit:
feat: HTTPMethod 열거형 구현 #3

Commit:
feat: Gettable Protocol 구현 #3

- APIProtocol 상속

Commit:
feat: URLRequest+extension Gettable 프로토콜 매개변수로 받는 init?() 추가 #3

Commit:
refactor: NetworkService request() 개선 #3

- URLRequest+extension init(api: Gettable) -> init(api: APIProtocol)

Commit:
feat: URLSession extension URLSessionProtocol 채택 #3

Commit:
test: MockURLSession 타입 구현 #3

- URL이 유효하지않을때 statusCodeError가 나오는지 Unit Test 진행

Commit:
test: URL이 Nil일때 urlIsNil에러가 나오는지 Unit Test 진행 #3

Commit:
test: MockURLSession의 isSuccess가 true일때 정상동작 하는지 Unit Test 진행 #3

Commit:
test: MockURLSession의 isSuccess가 false일때 실패하는지 Unit Test 진행 #3

Commit:
refactor: NetworkError isURLNil -> invalidURLRequest #3

Commit:
test: MockURLSession의 MockError을 담아보냈을때 MockError을 반환하는지 Unit Test 진행 #3

donggeonoh commented on BithumbYagomAcamedy/Extension/URLSession+extension.swift:
```diff
@@ -0,0 +1,16 @@
+
+import Foundation
+
+protocol URLSessionProtocol {
```
URLSessionProtocol 보단, URLSessionProviding 과 같은 네이밍이면 더 좋을 것 같습니다.

HJEHA replied:
의견 감사합니다:)

donggeonoh commented on BithumbYagomAcamedy/Extension/URLSession+extension.swift:
```diff
@@ -0,0 +1,16 @@
+    ) -> URLSessionDataTask
+}
+
+extension URLSession: URLSessionProtocol { }
```
혹시 이 부분은 어떤 의미가 있을까요?

HJEHA replied:
`URLSession`을 `URLSessionProtocol`을 채택하면 `NetworkService`의 session 프로퍼티의 타입을 `URLSessionProtocol`로 지정할 수 있습니다.
이점으로는 실제 네트워크에 의존하지 않는 테스트를 진행할 때 `NetworkService`의 session 프로퍼티를 `URLSessionProtocol`을 채택한`MockURLSession`으로 변경이 용의하다는 이점이 있습니다:)

donggeonoh replied:
감사합니다. session의 success나 fail response, Error를 임의적으로 반환해줄 수 있는걸까요?

HJEHA replied:
회의실에서 의견 나눔.

donggeonoh commented on BithumbYagomAcamedy/Protocol/APIProtocol.swift:
```diff
@@ -0,0 +1,26 @@
+
+import Foundation
+
+protocol APIProtocol {
+    var url: URL? { get }
+    var method: HTTPMethod { get }
+}
```
APIProtocol 보단, ing, able 과 같은 네이밍을 갖추면 좋겠습니다.

HJEHA replied:
의견 감사합니다:)
`APIable`라는 네이밍 괜찮으신가요??

donggeonoh replied:
저는 괜찮을 것 같습니다 ㅎㅎ
@z3rosmith 어떤 의견이세요?

donggeonoh commented on BithumbYagomAcamedy/Protocol/APIProtocol.swift:
```diff
@@ -0,0 +1,26 @@
+    var method: HTTPMethod { get }
+}
+
+protocol Gettable: APIProtocol { }
```
Gettable에서 url, method도 받으니 parameter를 받는건 어떻게 생각하세요?
혹시 제가 잘못 이해하고 있는걸까요? 이전에 설명을 듣긴 했지만, 죄송하지만 다시 설명해주시면 감사하겠습니다.

HJEHA replied:
파라미터를 어디서에 받는지 구체적으로 설명해주실 수 있으신가요???
API 통신(HTTP)할 때 `get`, `post`, `put` 등 HTTPMethod에서 공통적으로 가지는 프로퍼티라고 판단해 `APIProtocol`을 따로 구현했습니다.
또 `APIProtocol`을 따로 만들고 `Gettable`에 상속해준 이유는 `get` 메소드에서는 `url`, `HTTPMethod`만 필요하기 때문입니다.

추가적으로 `APIProtocol`에 `body` 등 다른 프로퍼티가 없는 이유는 SOLID 원칙에 인터페이스 분리 원칙(ISP), 개방 폐쇄 원칙(OCP)을 고려한 설계입니다.

`APIProtocol`에 `body`가 존재할 경우 `Gettable`는 사용하지 않는 인터페이스가 생기기 때문에 추후 `post`기능이 생길 경우 `Postable` 프로토콜의 `body`에 추가하는 것이 맞다는 판단을 했습니다.(인터페이스 분리 원칙 : ISP)
또, 위와 같은 설계를 가져갈 경우 추후 `post`기능이 생겼을 때 `Postable` 프로토콜만 추가해주면 되기 때문에 이전에 작성한 `APIProtocol`, `Gettable`을 수정하지 않아도 기능을 추가할 수 있다는 이점이 있습니다.(개방 폐쇄 원칙 : OCP)

donggeonoh replied:
아 gettable에서 parameter 프로퍼티를 추가하는게 어떤지 해서요 ㅎㅎ
답변 감사합니다.

donggeonoh commented on BithumbYagomAcamedy/Utility/NetworkService.swift:
```diff
@@ -0,0 +1,75 @@
+    case invalidURLRequest
+    case unknown(error: Error)
+
+    var errorDescription: String? {
+        switch self {
+        case .statusCodeError:
+            return "정상적인 StatusCode가 아닙니다."
+        case .invalidURLRequest:
+            return "정상적인 URLRequest가 아닙니다."
+        case .unknown(let error):
+            return "\(error.localizedDescription) 에러가 발생했습니다."
+        }
+    }
```
NameSpace와 같은 폴더를 하나 만들어서 문자열을 관리했으면 좋겠습니다.
저는 NetworkErrorNameSpace enum 타입 안에 static let으로 관리하는 방법을 사용했었는데요. 어떠신가요?

HJEHA replied:
문자열을 따로 관리하자는 의견 감사합니다:)
다만 열거형의 case를 이용할 수도 있는데 이전에 static let을 사용하신 이유를 알 수 있을까요???

donggeonoh replied:
만약 case를 사용하게 되면, 마지막에 rawValue가 붙기 때문에 static let으로 사용하였습니다~

donggeonoh commented on BithumbYagomAcamedy/Utility/NetworkService.swift:
```diff
@@ -0,0 +1,75 @@
+import Foundation
+
+enum NetworkError: LocalizedError, Equatable {
+    static func == (lhs: NetworkError, rhs: NetworkError) -> Bool {
+        return lhs.errorDescription == rhs.errorDescription
+    }
```
이 부분은 타입에 대해 같은지 재정의 하는 것인데, error 요약문으로 비교해도 괜찮은걸까요? 이유가 궁금하네요.

HJEHA replied:
unknown(error: Error) case 때문에 Unit Test를 작성할 때 error를 비교할 수 없어 Equatable을 채택하고 `errorDescription`으로 비교하는 방식을 취했습니다.
Unit Test에서 필요한 부분이니 테스트 코드에서 NetworkError의 extension을 활용해 Equatable 채택해주면 좋을 것 같네요:)
의견 감사합니다:)

donggeonoh commented on BithumbYagomAcamedy/Utility/NetworkService.swift:
```diff
@@ -0,0 +1,75 @@
+        return lhs.errorDescription == rhs.errorDescription
+    }
+    
+    case statusCodeError
```
statuscode도 같이 넘어가면 좋을 것 같습니다.

HJEHA replied:
좋은 의견입니다:)

donggeonoh commented on BithumbYagomAcamedy/Utility/NetworkService.swift:
```diff
@@ -0,0 +1,75 @@
+                return
+            }
+            
+            let successStatusCode = 200..<300
```
상숫값도 NamaSpace와 같은 하나의 타입으로 관리하는 것이 좋을 것 같습니다. 여기선 상수를 여기 한곳에만 쓰이긴 하지만, 한 곳에서 상수를 관리하는 것이 좋을 것 같아서요.

HJEHA replied:
상수나 변수를 따로 모아두면 사용하기 편하다는 이점은 있지만, 저는 조금 다르게 생각합니다.
전역적으로 사용하는 상수, 변수가 많아질 경우 개발자가 관리해야하는 변수, 상수가 많아지고 그렇게 되면 휴먼 에러가 발생할 가능성이 높아진다고 생각합니다. 또, 전역 변수의 경우 여러 코드(메서드 등)에서 접근할 수 있기 때문에 사이드 이펙트가 생길 가능성도 높아질 것 같습니다.

donggeonoh replied:
그렇다면, 멤버 프로퍼티로 정의하면 어떨까요?

donggeonoh commented on BithumbYagomAcamedyTests/NetworkServiceTest/Mock/MockURLSession.swift:
```diff
@@ -0,0 +1,69 @@
+            return URLSessionDataTask()
+        }
+        
+        let successResponse = HTTPURLResponse(
+            url: url,
+            statusCode: 200,
+            httpVersion: "1.1",
+            headerFields: nil
+        )
```
궁금해서 여쭤보겠습니다!!! httpVersion은 넣어줘야 하는 값일까요? (디폴트 값은 없는건가요?) 저희가 일반적으로 사용하고 있는 버전이 무엇인지도 궁금하네요.

HJEHA replied:
저도 같은 궁금점이 있어 작업할 때 공식문서를 찾아봤습니다.
공식 문서에서는 일반적으로 `1.1`을 사용한다고 적혀있어 `1.1`로 코드를 작성했습니다.
- httpVersion
    > 서버에 반환된 httpVersion 일반적으로 1.1을 사용함.
	```swift
	let successResponse = HTTPURLResponse(
	    url: url,
	    statusCode: 200,
	    httpVersion: "1.1",
	    headerFields: nil
	)
	```
    >[httpVersion 애플 공식 문서](https://developer.apple.com/documentation/foundation/httpurlresponse/1415870-init)
참고해보시면 좋을 것 같습니다.

donggeonoh commented on BithumbYagomAcamedyTests/NetworkServiceTest/NetworkServiceTests.swift:
```diff
@@ -0,0 +1,99 @@
+    var mockAPI: MockAPI!
+    
+    override func setUpWithError() throws {
+        try super.setUpWithError()
```
궁금해서 질문 드립니다
각 초기 설정 메서드에서 super를 불러주는 이유가 무엇인지 궁금합니다!

HJEHA replied:
오버라이드해서 사용하는 메서드이기도 하고 공식문서에 super에 대한 언급이 없어서 super를 호출해줬습니다.
다시 공식문서를 읽어보니 초기 설정할 때 에러가 발생할 수 있으면 setUpWithError을 재정의해서 사용하라고 적혀있네요.
또, super를 제거하고 Unit Test를 실행했을 때 문제도 발생하지 않았습니다.
다음 커밋때 반영하겠습니다.
의견 감사합니다:)

donggeonoh replied:
super 메서드를 호출하지 않아도 되는걸까요?
사실 필요에 의해서라면 삭제하지 않아도 된다고 생각합니다!
궁금해서 여쭤봤네요 ㅎㅎ

donggeonoh commented on BithumbYagomAcamedyTests/NetworkServiceTest/Mock/MockURLSession.swift:
```diff
@@ -0,0 +1,69 @@
+            headerFields: nil
+        )
+        
+        let dataString = #""OK""#
```
테스트 시 사용되는 것 같은데, session 클래스의 멤버로 갖고 있으면 어떨까요?

HJEHA replied:
이 경우도 위에 상수를 한곳에 두고 관리하면 어떻겠냐는 의견의 답변과 동일합니다.
함수 내부에서만 사용되는 상수인데 프로퍼티로 가지게되면 상수의 노출 범위가 너무 넓다고 생각합니다.
이 부분에 대해서는 팀원들끼리 회의를 진행하면 좋을 것 같습니다:)

donggeonoh replied:
클래스 멤버에서 접근하는 것이라서, 괜찮다고 생각하는데요.
어떤 의견이실까요? @z3rosmith

donggeonoh reviewed (changes requested):
Network 설계는 제가 예전에 구현했던 것 보다 더 잘 구현되어있다고 생각합니다.
네이밍이나 상수 관련한 부분만 조금 더 고민하면 좋을 것 같습니다!

z3rosmith commented on BithumbYagomAcamedy/Extension/URLSession+extension.swift:
```diff
@@ -0,0 +1,16 @@
+protocol URLSessionProtocol {
+    func dataTask(with request: URLRequest,
+                  completionHandler: @escaping (Data?, URLResponse?, Error?) -> Void
+    ) -> URLSessionDataTask
```
저희가 StyleShare의 convention을 따르기로 했는데 다음과 같이 바꾸는게 어떨까요?

- 함수 정의가 최대 길이를 초과하는 경우에는 아래와 같이 줄바꿈합니다.
```swift
func collectionView(
  _ collectionView: UICollectionView,
  cellForItemAt indexPath: IndexPath
) -> UICollectionViewCell {
  // doSomething()
}

func animationController(
  forPresented presented: UIViewController,
  presenting: UIViewController,
  source: UIViewController
) -> UIViewControllerAnimatedTransitioning? {
  // doSomething()
}
```

HJEHA replied:
줄바꿈이 어색하다고 느꼈었는데 파라미터 with에서 줄바꿈이 안되어 있었네요:)
피드백 감사합니다.

donggeonoh replied:
좋은 의견입니다. 동의합니다 ㅎㅎ

Commit:
refactor: 컨벤션 개선 #3

- URLSessionProviding dataTask() 줄바꿈 개선
- APIProtocol -> APIable 네이밍 변경
- URLSessionProtocol -> URLSessionProviding 네이밍 개선

Commit:
refactor: NetworkError Equatable 프로토콜 채택 위치 개선 #3

- Unit Test에서 Error를 비교하기 위한 프로토콜이므로 extension을 활용해 Unit Test 코드로 옮김

Commit:
refactor: NetworkError StatusCode 케이스의 연관값으로 statusCode 포함 #3

- 어떤 statusCode가 반환됐는지 명확하게 알수 있기 때문이다
- URL이_유효하지않을때_Error가_나오는지 개선 : 유효하지 않는 URL일 경우 error를
  시스템이 반환해 주기 때문에 어떤 error이 반환되는지 명확하게 알 수 없다.
  따라서 error가 발생했는지 파악하도록 개선했다
- XCTestExpectation() 사용 : URLSession은 비동기 작업이기 때문에 비동기 작업을
  테스트하기 위함이다

Commit:
Merge branch 'develop' into feat/3/Network
